### PR TITLE
Add square, cube and threshold cli

### LINKED ISSUE #88: Create CLI for creating mask
## Context

Command Line Interface allow to use the python software on the command line. Currently, the masking methods implemented are [threshold](https://github.com/shimming-toolbox/shimming-toolbox-py/blob/master/shimmingtoolbox/masking/threshold.py) and [shape](https://github.com/shimming-toolbox/shimming-toolbox-py/blob/master/shimmingtoolbox/masking/shapes.py) where shape can be a square or a cube. The goal is to implement a basic CLI that enables to call the different implemented masking methods while considering what will be implemented in the future (SCT, BET)

## Discussion
Suggestion for name:
- st_mask

Suggestions for flags:

~~~
-i <file> input
-method {sct, bet, shape, threshold}
-shape {cube, ball} Only for method=shape
-size <int> Radius of the shape in voxel (not mm). Only for method=shape
~~~

Potential new feature: have it in mm. To discuss...

## To get started
Where: [shimmingtoolbox/cli](https://github.com/shimming-toolbox/shimming-toolbox-py/tree/master/shimmingtoolbox/cli)
Example: [shimmingtoolbox/cli/download_data](https://github.com/shimming-toolbox/shimming-toolbox-py/blob/master/shimmingtoolbox/cli/download_data.py)
Useful link to start [contributing](https://shimming-toolbox.org/en/latest/3_contributing/CONTRIBUTING.html)

### PR DESCRIPTION
**Pull request by maheuss**
Create CLI for creating mask

Implementation of st_mask_cube, st_mask_threshold and st_mask_square which allow to apply these masks to the desired files. Each function has its own required input sizes.

Fixes #88
Related to #66

### CODE REVIEW

File: shimmingtoolbox/cli/st_mask_square.py
Comment (po09i) [inline shimmingtoolbox/cli/st_mask_square.py]:
```suggestion
def st_mask_square(filename, len_dim1, len_dim2, center_dim):
```

File: shimmingtoolbox/cli/st_mask_square.py
Comment (po09i) [inline shimmingtoolbox/cli/st_mask_square.py]: Initialize to the correct size. You can use `np.zeros_like` I think

File: shimmingtoolbox/cli/st_mask_square.py
Comment (po09i) [inline shimmingtoolbox/cli/st_mask_square.py]: Square will mostly be used for 2d images, so I would make sure it works for 3d but this will crash for arrays with only 2 dimensions since shape[2] will not exist

File: shimmingtoolbox/cli/st_mask_cube.py
Comment (po09i) [inline shimmingtoolbox/cli/st_mask_cube.py]: You can look at the wiki for variable name conventions. In this case, it should be `nii_data`

File: setup.py
Comment (jcohenadad) [inline setup.py]: don't we want to instead deal with a single CLI called `st_mask`, and input the shape as param?

File: setup.py
Comment (po09i) [inline setup.py]: Wrong PR?

File: setup.py
Comment (maheuss) [reply setup.py]: Feasible with the click.group command

File: setup.py
Comment (po09i) [inline setup.py]:
```suggestion
            "st_mask=shimmingtoolbox.cli.mask:main",
```

The name of the file should change to fit the new name.

File: shimmingtoolbox/cli/mask_shimmingtoolbox.py
Comment (po09i) [inline shimmingtoolbox/cli/mask_shimmingtoolbox.py]: Can we add the same context_settings to be able to display -h

File: test/cli/test_cli_mask_shimmingtoolbox.py
Comment (po09i) [inline test/cli/test_cli_mask_shimmingtoolbox.py]: I think this is bypassing the click.group. I'm thinking importing main (which could be called mask) and the particular  tests would need the correct argument ('cube', ...)

**Review by po09i — COMMENTED** [review]
Just a couple of suggestions to make things more uniform.

- Name of CLI should not have shimmingtoolbox in the name. --> st_mask
- The main could be the same name as the file (mask).

Output: The CLI should save the mask as a nifti array using an option spcifying where to put the output file. The fullpath should be returned.

**Comment by po09i** [comment]
LGTM!

File: shimmingtoolbox/cli/mask.py
Comment (jcohenadad) [inline shimmingtoolbox/cli/mask.py]:
```suggestion
@click.group(context_settings=CONTEXT_SETTINGS, help=f"Create a mask based on a specified shape (cube, square, etc.) or based on the thresholding of an input image.")
```

File: shimmingtoolbox/cli/mask.py
Comment (jcohenadad) [inline shimmingtoolbox/cli/mask.py]: in a cube, the three dimensions are the same, hence you only one flag

File: shimmingtoolbox/cli/mask.py
Comment (jcohenadad) [inline shimmingtoolbox/cli/mask.py]: it's heavy to have to enter those dims separately. you could have a simpler approach:
~~~
st_mask -center x y z
~~~

File: shimmingtoolbox/cli/mask.py
Comment (jcohenadad) [reply shimmingtoolbox/cli/mask.py]: same comment as in https://github.com/shimming-toolbox/shimming-toolbox/pull/137/files#r519030186, lighter code and more intuitive to have (replacing len by size):
~~~
st_mask size x
~~~

also: avoid mixing - and _

File: shimmingtoolbox/cli/mask.py
Comment (jcohenadad) [inline shimmingtoolbox/cli/mask.py]: should be checked by the parser

Comment (jcohenadad) [reply shimmingtoolbox/cli/mask.py]: argparse can check for that-- i'm sure click can do it (otherwise that's a strong cons against click)

File: shimmingtoolbox/cli/mask.py
Comment (jidicula) [inline shimmingtoolbox/cli/mask.py]: good use of comments

File: shimmingtoolbox/cli/mask.py
Comment (jidicula) [inline shimmingtoolbox/cli/mask.py]: I'm always happy to see f-strings!

**Review by jidicula — CHANGES_REQUESTED** [review]
Looks good! Just one small change requested so we can keep our docs up-to-date with the library code:

Could you add this CLI to the documentation? An example can be seen [here](https://github.com/shimming-toolbox/shimming-toolbox/blame/master/docs/source/6_api_reference/api.rst#L51) for `download_data`. If you want to check that it builds correctly with Sphinx, you can `cd` to `docs/` and run `make html`, then check `shimming-toolbox/docs/build/html/6_api_reference/api.html` in any web browser. If you're not able to build the docs locally, you can push your .rst changes anyway and check the branch preview on ReadTheDocs: 
Screen Shot 2020-11-08 at 14 27 05

**Review by jcohenadad — CHANGES_REQUESTED** [review]
I left a few comments that need to be addressed before being able to merge

File: shimmingtoolbox/cli/mask.py
Comment (maheuss) [reply shimmingtoolbox/cli/mask.py]: Is it the same principle for the square mask or do we have to consider 2 sizes for this one?

Comment (jcohenadad) [reply shimmingtoolbox/cli/mask.py]: same principle

File: shimmingtoolbox/cli/mask.py
Comment (jidicula) [inline shimmingtoolbox/cli/mask.py]: I think this should be `@click.command()` instead of `@mask.command()` 😅 
```suggestion
@click.command(context_settings=CONTEXT_SETTINGS,
```

Comment (jidicula) [reply shimmingtoolbox/cli/mask.py]: (this is what was causing the RTD build to fail)

**Review by jidicula — CHANGES_REQUESTED** [review]
Small change requested to fix a decorator typo. Once fixed, RTD build should succeed.

File: shimmingtoolbox/cli/mask.py
Comment (maheuss) [reply shimmingtoolbox/cli/mask.py]: Knowing that I use click.group(), each subcommand must be implemented with name_main_function.command(). In this case, the parent function is mask so each subcommand is implemented with mask.command(). I tried multiple methods to document all these functions with automodule but none of them managed to display mask and its subcommands. Otherwise, there is a sphinx extension specially designed for the click library (sphinx_click.ext) and I managed to generate the documentation with it. The problem with this alternative is that the layout is not exactly the same as with automodule.

File: docs/source/conf.py
Comment (po09i) [inline docs/source/conf.py:36]: sphinx_click needs to be installed with pypi. So you need to add it to the `setup` file within extra_requires: docs.

File: docs/source/6_api_reference/api.rst
Comment (po09i) [inline docs/source/6_api_reference/api.rst]: `mask_cli` does not exist, it is named `mask`.

File: docs/source/conf.py
Comment (maheuss) [reply docs/source/conf.py:36]: Added now !

File: docs/source/6_api_reference/api.rst
Comment (maheuss) [reply docs/source/6_api_reference/api.rst]: The name of my function in mask.py is now changed to mask_cli

File: shimmingtoolbox/cli/mask.py
Comment (maheuss) [reply shimmingtoolbox/cli/mask.py]: I adapted it this way, thank you!

File: shimmingtoolbox/cli/mask.py
Comment (maheuss) [reply shimmingtoolbox/cli/mask.py]: I removed it as it was not necessary

File: shimmingtoolbox/cli/mask.py
Comment (jidicula) [inline shimmingtoolbox/cli/mask.py]: @jcohenadad I wonder if the triple-quoted docstrings for click functions become redundant now that we use `sphinx_click` extension. can we forego those docstrings for any click functions and document them through these decorators instead?

**Review by jidicula — APPROVED** [review]
Travis and RTD builds look good to me - good call with using the sphinx_click extension. This makes much more sense for autogenerating documentation of the CLI modules going forward.

I'll leave it to others to review whether the actual mask generation UX is sound - it's outside my expertise 😅

**Review by jcohenadad — CHANGES_REQUESTED** [review]
The CLI doc is not very helpful:
image

File: test/cli/test_cli_mask.py
Comment (jcohenadad) [inline test/cli/test_cli_mask.py]: it would be nice to check that the expected mask is created

File: shimmingtoolbox/cli/mask.py
Comment (jcohenadad) [inline shimmingtoolbox/cli/mask.py]: every edge of a cube have the same length, so no need to mention the 1st 2nd and 3rd dim.
also, use single quote to be consistent (and as per guidelines)
```suggestion
@click.option('-size', required=True, type=int, help="Edge size of the cube (in pixels)")
```

File: shimmingtoolbox/cli/mask.py
Comment (jcohenadad) [inline shimmingtoolbox/cli/mask.py]:
```suggestion
@click.option('-center', nargs=3, type=int, default=(None, None, None), help="Center of the cube along first, second "
```

File: shimmingtoolbox/cli/mask.py
Comment (jcohenadad) [inline shimmingtoolbox/cli/mask.py]: is this working if center is none?

File: shimmingtoolbox/cli/mask.py
Comment (jcohenadad) [inline shimmingtoolbox/cli/mask.py:55]: why not accepting 4d data? e.g. our temporal fieldmaps used for realtime shimming.
what do you think @po09i ?

File: shimmingtoolbox/cli/mask.py
Comment (jcohenadad) [inline shimmingtoolbox/cli/mask.py]: i thought that with the new sphinx_click, these three-quotes were no more required?

File: shimmingtoolbox/cli/mask.py
Comment (jcohenadad) [inline shimmingtoolbox/cli/mask.py]: apply my comments raised for the cube case

File: shimmingtoolbox/cli/mask.py
Comment (jcohenadad) [inline shimmingtoolbox/cli/mask.py]: this behaviour should be described in the usage

File: shimmingtoolbox/cli/mask.py
Comment (jcohenadad) [inline shimmingtoolbox/cli/mask.py]:
```suggestion
@click.option('-thr', default=30, help="Value to threshold the data: voxels will be set to zero if their value is "
```

File: shimmingtoolbox/cli/mask.py
Comment (po09i) [reply shimmingtoolbox/cli/mask.py]: The mask is based off the API named cube which takes all 3 dimensions as inputs, the API could be named prism or rectangle to have a more appropriate name but to be consistent with the API I would leave it as is and revisit in another PR.

File: shimmingtoolbox/cli/mask.py
Comment (po09i) [reply shimmingtoolbox/cli/mask.py:55]: Indeed, it would be useful for masking the temporal data, I think it makes more sense to implement the 4d capability in the masking API rather than in the CLI.

File: shimmingtoolbox/cli/mask.py
Comment (jcohenadad) [reply shimmingtoolbox/cli/mask.py]: i see. in that case i would not call it “cube” because this is wrong. How about “box”?

and “rect” for the 2d version

File: shimmingtoolbox/cli/mask.py
Comment (maheuss) [reply shimmingtoolbox/cli/mask.py:55]: The shape_square and shape_cube functions are used on 2D and 3D arrays respectively, that's why I limited the use of CLI to 3D nifti files. I could extend and adapt the CLI to 4D arrays but I don't know if this would be optimal compared to adapting the masking API directly.

File: shimmingtoolbox/cli/mask.py
Comment (maheuss) [reply shimmingtoolbox/cli/mask.py]: The shape_cube function indicates that if the center (for a certain dimension) is None, then the middle of the dimension is used by default. The center is None for all 3 dimensions in my test_cli_mask_cube API and I got no error.

File: shimmingtoolbox/cli/mask.py
Comment (maheuss) [reply shimmingtoolbox/cli/mask.py]: Ok I will change the name of cube and square in this case!

File: shimmingtoolbox/cli/mask.py
Comment (jcohenadad) [reply shimmingtoolbox/cli/mask.py:55]: adapting the masking API would probably be better-- if everyone agrees pls open an issue and cross-reference it here so we don,t forget-- thanks!

File: shimmingtoolbox/cli/mask.py
Comment (maheuss) [reply shimmingtoolbox/cli/mask.py]: True, it's useless. I will remove it and be more precise in my decorators for a better documentation.

File: shimmingtoolbox/cli/mask.py
Comment (maheuss) [reply shimmingtoolbox/cli/mask.py]: Done now !

File: test/cli/test_cli_mask.py
Comment (maheuss) [reply test/cli/test_cli_mask.py]: I'm going to think about how to add this check

File: docs/source/6_api_reference/api.rst
Comment (jidicula) [inline docs/source/6_api_reference/api.rst]: We have a new section for CLI documentation [here](https://github.com/shimming-toolbox/shimming-toolbox/blob/master/docs/source/6_cli_reference/cli.rst) - could you move these stubs into there?

**Review by jidicula — CHANGES_REQUESTED** [review]

File: docs/source/6_api_reference/api.rst
Comment (maheuss) [reply docs/source/6_api_reference/api.rst]: Done now, thank you for the information !

File: shimmingtoolbox/cli/mask.py
Comment (po09i) [inline shimmingtoolbox/cli/mask.py]: Le output devrait être un nom de fichier (comme ça il sera possible d'appeler mask 2 fois de suite sans avoir un problèeme de duplication du nom de fichier) le default pourrait être: `default=os.path.join(os.curdir, 'mask.nii.gz')`. N'oublie pas de mentionné dans le help les extensions supportés soit `.nii` et `.nii.gz`.

File: test/cli/test_cli_mask.py
Comment (po09i) [reply test/cli/test_cli_mask.py]: You can open it and make sure some pixels are what you are expecting

**Review by jcohenadad — APPROVED** [review]

**Comment by po09i** [comment]
Remaining work to do summarized in this [issue](https://github.com/shimming-toolbox/shimming-toolbox/issues/192). This can be merged.

**Review by jidicula — APPROVED** [review]
LGTM 🚢

File: shimmingtoolbox/cli/mask.py
Comment (po09i) [reply shimmingtoolbox/cli/mask.py:55]: #193 opened

**Review by po09i — APPROVED** [review]

**Pull request merged**
